### DISCUSSION TITLE
Boundary condition that couples Neumann data to the time derivative of a variable

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#20850] [Q&A General] [answered]
Hi

I need to define the following boundary condition:
`du/dx=coeff * du/dt`

Correct me if I am wrong, `CoupledVarNeumannBC` will not be useful in this case.

Any ideas!

**Comment by GiudGiud** [comment, marked as answer]
Hello

You could still use `CoupledVarNeumannBC`, if the coupled variable was an auxvariable equal to the derivative. I dont see an auxkernel to do that but it d be pretty easy. Please let know if you want to do that

Guillaume

**Reply by aaelmeli** [reply]
> Hello
> 
> You could still use `CoupledVarNeumannBC`, if the coupled variable was an auxvariable equal to the derivative. I dont see an auxkernel to do that but it d be pretty easy. Please let know if you want to do that
> 
> Guillaume

Thanks, @GiudGiud. Yes, that would be great.

**Reply by GiudGiud** [reply]
Ok I ll make an object today and make a PR to MOOSE

**Reply by GiudGiud** [reply]
Could you please try out the code in https://github.com/idaholab/moose/pull/20853
and see if it works for you.

**Reply by aaelmeli** [reply]
> Could you please try out the code in #20853 and see if it works for you.

Hi @GiudGiud 
What command should I use to pull this? I think I am doing something wrong when pulling this to my local.

**Reply by GiudGiud** [reply]
There's multiple ways. One would be
- add my branch as remote
- fetch it
- add the two commits from my PR

```
git remote add guillaume git@github.com:GiudGiud/moose.git
git fetch guillaume
git cherry-pick 3f51a3271b1e7c2637aa4e7607118955bf5b66e1
git cherry-pick b233d0b0a77ade27c811d550d274f7891da2f663
```

**Reply by aaelmeli** [reply]
Thank you, but I am still getting this error. 
```
(moose) aaelmeli@CCEE-DT-284:~/projects/moose$ git remote add guillaume git@github.com:GiudGiud/moose.git
(moose) aaelmeli@CCEE-DT-284:~/projects/moose$ git fetch guillaume
Warning: Permanently added the ECDSA host key for IP address '140.82.112.3' to the list of known hosts.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Reply by aaelmeli** [reply]
![image](https://user-images.githubusercontent.com/88537645/164950987-a525b0c1-aa71-4871-8b04-c34670cb1745.png)
This is a screenshot of the branches that I have. I think I mistakenly made the `moose.git` branch with some command. I do not know what is the significance of the `"+" `sign on its left. This might be the problem.

I tried to delete it but I am getting the following error message:
`error: Cannot delete branch 'moose.git' checked out at '/home/aaelmeli/projects/moose/git@github.com:GiudGiud/moose.git'`

**Reply by GiudGiud** [reply]
never seen that
what does `git remote -v` return?

did you clone MOOSE with ssh or with https?

**Reply by aaelmeli** [reply]
> never seen that what does `git remote -v` return?

```
abdorepo        https://github.com/aaelmeli/moose.git (fetch)
abdorepo        https://github.com/aaelmeli/moose.git (push)
github-desktop-giudgiud https://github.com/GiudGiud/moose.git (fetch)
github-desktop-giudgiud https://github.com/GiudGiud/moose.git (push)
guid    git@github.com:GiudGiud/moose.git (fetch)
guid    git@github.com:GiudGiud/moose.git (push)
guillaume       git@github.com:GiudGiud/moose.git (fetch)
guillaume       git@github.com:GiudGiud/moose.git (push)
origin  https://github.com/idaholab/moose.git (fetch)
origin  https://github.com/aaelmeli/moose_remote_aaelmeli.git (push)
origin2 git@github.com:aaelmeli/moose_remote_aaelmeli.git (fetch)
origin2 git@github.com:aaelmeli/moose_remote_aaelmeli.git (push)
upstream        https://github.com/idaholab/moose.git (fetch)
upstream        https://github.com/idaholab/moose.git (push)
```

> did you clone MOOSE with ssh or with https?

I have tried both actually with no success.

I could get you pr with the following command

`git fetch https://github.com/GiudGiud/moose.git PR_timeaux:timeaux`
updated moose and conda, and I am getting another interesting error when do `make `as follows:

```
(moose) aaelmeli@CCEE-DT-284:~/projects/first_test$ make
Rebuilding symlinks in /home/aaelmeli/projects/first_test/build/header_symlinks
Compiling C++ (in opt mode) /home/aaelmeli/projects/moose/framework/build/unity_src/actions_Unity.C...
In file included from /home/aaelmeli/projects/moose/framework/build/header_symlinks/MaterialBase.h:34,
                 from /home/aaelmeli/projects/moose/framework/build/header_symlinks/Material.h:13,
                 from /home/aaelmeli/projects/moose/framework/src/actions/CheckOutputAction.C:12,
                 from /home/aaelmeli/projects/moose/framework/build/unity_src/actions_Unity.C:49:
/home/aaelmeli/projects/moose/framework/build/header_symlinks/Assembly.h:24:10: fatal error: libmesh/elem_side_builder.h: No such file or directory
   24 | #include "libmesh/elem_side_builder.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [/home/aaelmeli/projects/moose/framework/build.mk:145: /home/aaelmeli/projects/moose/framework/build/unity_src/actions_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```

**Reply by GiudGiud** [reply]
so to make the ssh cloning work you ll have to go to Github settings and add your ssh public key from your machine. 

This does look like libmesh is out of date.
What does `mamba list | grep moose` return?

**Reply by lindsayad** [reply]
Retrieving the time derivative of a nonlinear variable through an auxiliary variable loses the Jacobian information. So if you go with the aux variable route, you will likely see a degradation in solver convergence

**Reply by aaelmeli** [reply]
> so to make the ssh cloning work you ll have to go to Github settings and add your ssh public key from your machine.
> 
> This does look like libmesh is out of date. What does `mamba list | grep moose` return?

Yes, I think the problem was in mamba was not there. I updated it and it worked fine.



> Retrieving the time derivative of a nonlinear variable through an auxiliary variable loses the Jacobian information. So if you go with the aux variable route, you will likely see a degradation in solver convergence

Yes, that is what I have noticed. Thanks for the note.
